macOS 11.5.2 build fails because of OSSpinLock deprecation

**Describe the bug**
A clear and concise description of what the bug is.

**To Reproduce**
Build using `make mac_osx` on Big Sur 11.5.2

**Expected behavior**
A clean build

**Hardware and operating system**
CPU: 2 GHz Quad-Core Intel Core i7
Memory: 8 GB 1600 MHz DDR3
Form factor: laptop
CPU arch: x86
Mac OS X: 11.5.2

**Additional context**

```
In file included from cli/main.c:1:
include/compiler.h:78:28: warning: 'OSSpinLock' is deprecated: first deprecated in macOS 10.12 - Use os_unfair_lock() from <os/lock.h> instead [-Wdeprecated-declarations]
    void pthread_spin_init(OSSpinLock * a_lock, void * a_opts);
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/libkern/OSSpinLockDeprecated.h:79:17: note: 'OSSpinLock' has been explicitly marked deprecated here
typedef int32_t OSSpinLock OSSPINLOCK_DEPRECATED_REPLACE_WITH(os_unfair_lock);
                ^
2 warnings generated.
make[2]: *** [engine] Error 1
make[1]: *** [mac_osx] Error 2
make: *** [mac_osx] Error 2
```
If someone hasn't already I can start a PR for this.

--- a/src/engine/Makefile
+++ b/src/engine/Makefile
@@ -18,7 +18,6 @@ CACHE_LINE_SIZE = $(shell getconf LEVEL1_DCACHE_LINESIZE || echo 64)
 OPTIMIZE_FLAGS ?= \
 	-ffast-math \
 	-finline-functions \
-	-floop-optimize \
 	-flto \
 	-fno-stack-protector \
 	-fprefetch-loop-arrays \

--- a/src/engine/include/compiler.h
+++ b/src/engine/include/compiler.h
@@ -70,12 +70,13 @@ struct SamplePair {
 #if SG_OS == _OS_MAC_OS_X
 
     #include <libkern/OSAtomic.h>
+    #include <os/lock.h>
 
-    #define pthread_spinlock_t OSSpinLock
-    #define pthread_spin_lock OSSpinLockLock
-    #define pthread_spin_unlock OSSpinLockUnlock
+    #define pthread_spinlock_t os_unfair_lock
+    #define pthread_spin_lock os_unfair_lock_lock
+    #define pthread_spin_unlock os_unfair_lock_unlock
 
-    void pthread_spin_init(OSSpinLock * a_lock, void * a_opts);
+    void pthread_spin_init(os_unfair_lock * a_lock, void * a_opts);
 
 #endif
 

--- a/src/engine/src/_main.c
+++ b/src/engine/src/_main.c
@@ -390,7 +390,7 @@ NO_OPTIMIZATION void destruct_signal_handling(){
 
 
 NO_OPTIMIZATION void set_thread_params(){
-#if SG_OS != _OS_LINUX && SG_OS != _OS_MAC_OS_X
+#if SG_OS != _OS_LINUX || SG_OS != _OS_MAC_OS_X
     log_info("Setting thread params");
     int f_current_proc_sched = sched_getscheduler(0);
 

--- a/src/engine/src/_main.c
+++ b/src/engine/src/_main.c
@@ -390,7 +390,7 @@ NO_OPTIMIZATION void destruct_signal_handling(){
 
 
 NO_OPTIMIZATION void set_thread_params(){
-#if SG_OS != _OS_LINUX
+#if SG_OS != _OS_LINUX && SG_OS != _OS_MAC_OS_X
     log_info("Setting thread params");
     int f_current_proc_sched = sched_getscheduler(0);
 

--- a/src/engine/src/_main.c
+++ b/src/engine/src/_main.c
@@ -390,7 +390,7 @@ NO_OPTIMIZATION void destruct_signal_handling(){
 
 
 NO_OPTIMIZATION void set_thread_params(){
-#if SG_OS != _OS_LINUX || SG_OS != _OS_MAC_OS_X
+#if SG_OS != _OS_LINUX && SG_OS != _OS_MAC_OS_X
     log_info("Setting thread params");
     int f_current_proc_sched = sched_getscheduler(0);
 

--- a/src/engine/src/compiler.c
+++ b/src/engine/src/compiler.c
@@ -14,8 +14,9 @@
 
 
 #if SG_OS == _OS_MAC_OS_X
-    void pthread_spin_init(OSSpinLock * a_lock, void * a_opts){
-        *a_lock = 0;
+    #include <os/lock.h>
+    void pthread_spin_init(os_unfair_lock * a_lock, void * a_opts){
+        #define OS_UNFAIR_LOCK_INIT
     }
 #endif
 

--- a/src/engine/src/ipc/unix.c
+++ b/src/engine/src/ipc/unix.c
@@ -49,6 +49,7 @@ void ipc_client_send(
     int n;
     char buffer[1024];
 
+#if SG_OS != _OS_MAC_OS_X
     sendto(
         SOCKET_DATA.sockfd,
         (const char*)message,
@@ -65,6 +66,26 @@ void ipc_client_send(
         (struct sockaddr*)&SOCKET_DATA.servaddr,
         &SOCKET_DATA.len
     );
+
+#else
+#define MSG_OOB  0x1
+    sendto(
+        SOCKET_DATA.sockfd,
+        (const char*)message,
+        strlen(message),
+        MSG_OOB,
+        (const struct sockaddr*)&SOCKET_DATA.servaddr,
+        sizeof(SOCKET_DATA.servaddr)
+    );
+    n = recvfrom(
+        SOCKET_DATA.sockfd,
+        (char*)buffer,
+        1024,
+        MSG_WAITALL,
+        (struct sockaddr*)&SOCKET_DATA.servaddr,
+        &SOCKET_DATA.len
+    );
+#endif
     buffer[n] = '\0';
 }
 
@@ -151,6 +172,7 @@ void* ipc_server_thread(void* _arg){
             engine_message.key,
             engine_message.value
         );
+#if SG_OS != _OS_MAC_OS_X
         sendto(
             sockfd,
             (const char*)response,
@@ -158,6 +180,16 @@ void* ipc_server_thread(void* _arg){
             MSG_CONFIRM,
             (const struct sockaddr*)&cliaddr,
             len
+        );
+#else
+        sendto(
+            sockfd,
+            (const char*)response,
+            response_len,
+            MSG_OOB,
+            (const struct sockaddr*)&cliaddr,
+            len
+#endif
         );
     }
     return 0;

--- a/src/setup.py
+++ b/src/setup.py
@@ -42,7 +42,7 @@ else:
 OPT_FLAGS = ENV['OPT_FLAGS'].split() if 'OPT_FLAGS' in ENV else [
 	'-ffast-math',
 	'-finline-functions',
-	'-floop-optimize',
+#	'-floop-optimize',
 	'-flto',
 	'-fno-stack-protector',
 	'-fprefetch-loop-arrays',

--- a/src/sglib/lib/util.py
+++ b/src/sglib/lib/util.py
@@ -13,7 +13,7 @@ import sys
 import time
 
 
-if IS_LINUX:
+if IS_LINUX or IS_MAC_OSX:
     from .path.linux import *
 elif IS_WINDOWS:
     from .path.windows import *

